### ISSUE TITLE
CA1003: False positive for EventHandler<T> without generic constraint

### ISSUE THREAD

**Issue by SebastianSchumann** [dotnet/roslyn-analyzers#3862]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.0.0 (Latest)

#### Diagnostic ID

[CA1003](https://docs.microsoft.com/visualstudio/code-quality/ca1003)

#### Repro steps

    public class Sample<T>
    {
        public event EventHandler<T> XXX;
    }

There is no contraint for `T` as the `EventHandler<T>` does not enforce any contraint.

#### Expected behavior

No diagnostics are reported because there is no need to derive from `EventArgs`.

#### Actual behavior

The reported message is `Change the event 'XXX' to replace the type 'System.EventHandler<T>' with a generic EventHandler, for e.g. EventHandler<T>, where T is a valid EventArgs.`

**Comment by Evangelink** [comment]
@mavasani I am not totally sure to understand the extent of the rule. Reading the documentation, I would expect any code already defining the type `EventHandler` whatever the generic type passed to not raise a diagnostic.

BTW I confirm that the provide sample reproduce the issue.

**Comment by mavasani** [comment]
Seems like a bug in implementation.

**Comment by Evangelink** [comment]
I am just reviewing this issue and I confirm that I can reproduce the issue being raised. Yet I am not sure that the issue is wrong, I mean the message clearly needs to be tweaked and also maybe the doc.

At first, and reading the doc, I was expecting the rule to only report on where the type is a delegate which isn't `EventHandler<T>` but looking at some of the [tests](https://github.com/dotnet/roslyn-analyzers/blob/master/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstancesTests.cs#L41-L80) it seems that the analyzer also ensures that the generic type inside the `EventHandler` inherits `EventArgs`. If that's what we want to do then this is not a False Positive.

@mavasani Could you have a look and let me know what is supposed to be the scope of this rule?

**Comment by weezlemsneezle** [comment]
I am using `net5.0` and `<AnalysisMode>AllEnabledByDefault</AnalysisMode>`.
With the following code I get a CA1003
```
public struct Data
{
    public int Data1;
    public int Data2;
}

public class Receiver
{

    public delegate void DataReceivedHandler (Receiver sender, in Data data);

    public event DataReceivedHandler DataReceived;
}
```
After reading the doc this is rather unexpected, especially because I can't declare an `in` parameter with `EventHandler<>`

**Comment by Evangelink** [comment]
Ping @mavasani

**Comment by Rabadash8820** [comment]
Just ran into this issue myself. Declaring an event with `EventHandler<T>` where `T` is is not derived from `EventArgs` gets flagged for CA1003. Before I turned on the .NET5 analyzers, I had no reason to question declarations like this since they compiled just fine. I get that the whole point of analyzer rules is to catch these corner cases where the code _compiles_ but isn't necessarily following best practices. However, I'm not convinced that the occasional `EventHandler<int>` or `EventHandler<string>` is evil; I find it much cleaner than declaring a new class with only one member that derives from `EventArgs`. If the original authors of `EventHandler<T>` saw fit not to add a `where T : EventArgs` generic constraint, then that seems like reason enough to me for CA1003 to permit such code. @weezlemsneezle's use case is also worth ignoring.

**Comment by SebastianSchumann** [comment]
@Rabadash8820 Not even that the original authors did not add a Contraint to `EventHandler<T>` - they actively removed that constraint:
.Net 4.0 having such constraint:
https://docs.microsoft.com/en-us/dotnet/api/system.eventhandler-1?view=netframework-4.0
whereas in .Net 4.5+ it is removed:
https://docs.microsoft.com/en-us/dotnet/api/system.eventhandler-1?view=netframework-4.5

For me that's the main reason that this rule should not be reported!
